My prediction is that Random Forest Classifier will be better for this dataset because it can handle outliers better than a logistic regression model.

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [7]:
# Import the data
Credit_Risk_df= pd.read_csv("C:/Users/prakb/OneDrive/Desktop/supervised-machine-learning-challenge/Resources/lending_data.csv")
Credit_Risk_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

In [9]:
# Stastical analysis for dataset.
Credit_Risk_df.describe()


loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt   loan_status  
count     77536.000000      77536.000000  77536.000000  77536.000000  
mean          3.826610          0.392308  19221.949804      0.032243  
std           1.904426          0.582086   8371.635077      0.176646  
min           0.000000          0.000000      0.000000      0.000000  
25%           3.000000          0.000000  14800.000000      0.000000  
50%           4.000000          0.000000  18100.000000      0.000000  
75%           4.000000          1.000000  21400.000000      0.000000  
max          16.000000          3.000000  75200.000000      1.000000

From the dataset you can see that the mean is 9500 for loan size however the max value is 23,000. This tells us there are outliers in this dataset which would support my prediction because the Logitical Regression Model does not handle outliers in the data as well at Random Forest Classifier.

In [18]:
# Split the data into X_train, X_test, y_train, y_test
x =Credit_Risk_df.drop("loan_status", axis=1)
y = Credit_Risk_df["loan_status"] !=0

x_train, x_test, y_train, y_test = train_test_split(x, y, random_state=1)
scaler = StandardScaler().fit(x_train)
x_train

loan_size  interest_rate  borrower_income  debt_to_income  \
29175     8600.0          6.792            44500        0.325843   
23020     7800.0          6.419            41000        0.268293   
31269    10000.0          7.386            50100        0.401198   
35479     9300.0          7.093            47300        0.365751   
13470     9200.0          7.045            46900        0.360341   
...          ...            ...              ...             ...   
20609     7200.0          6.177            38700        0.224806   
21440    10000.0          7.389            50100        0.401198   
73349    10200.0          7.463            50800        0.409449   
50057    11100.0          7.838            54400        0.448529   
5192     10600.0          7.632            52400        0.427481   

       num_of_accounts  derogatory_marks  total_debt  
29175                3                 0       14500  
23020                2                 0       11000  
31269                4                 1       20100  
35479                3                 0       17300  
13470                3                 0       16900  
...                ...               ...         ...  
20609                1                 0        8700  
21440                4                 1       20100  
73349                4                 1       20800  
50057                5                 1       24400  
5192                 5                 1       22400  

[58152 rows x 7 columns]

In [20]:
x_train2 = scaler.transform(x_train)
x_test2 = scaler.transform(x_test)
x_train2

array([[-0.57708952, -0.56367666, -0.5652314 , ..., -0.43489843,
        -0.67289855, -0.5652314 ],
       [-0.95927354, -0.98302549, -0.98332378, ..., -0.96014741,
        -0.67289855, -0.98332378],
       [ 0.09173251,  0.10413354,  0.10371642, ...,  0.09035056,
         1.04334691,  0.10371642],
       ...,
       [ 0.18727852,  0.19070153,  0.18733489, ...,  0.09035056,
         1.04334691,  0.18733489],
       [ 0.61723554,  0.61229888,  0.61737277, ...,  0.61559954,
         1.04334691,  0.61737277],
       [ 0.37837052,  0.3807014 ,  0.37846284, ...,  0.61559954,
         1.04334691,  0.37846284]])

In [21]:
# Train a Logistic Regression model print the model score
logisitc_regmodel = LogisticRegression().fit(x_train, y_train)

In [22]:
logisitc_regmodel.score(x_train, y_train)

0.9921240885954051

In [24]:
# Train a Random Forest Classifier model and print the model score
random_forestclass = RandomForestClassifier(random_state=1, n_estimators=500).fit(x_train, y_train)

In [25]:
random_forestclass.score(x_train,y_train)

0.9975409272252029

My predicition was correct. Random forest Classifier is the better model for this dataset as it gave a 0.9975 accuracy compared to the 0.9921 accuracy given by Logistic Regression Model. Despite both models being very close in their accuracy I believe the Random Forest Classifier is higher because it can handle outliers better and as we saw above there are outliers in this dataset.